File: LeetCode/String/Medium/3. Longest Substring Without Repeating Characters/main.py
class Solution:
    def lengthOfLongestSubstring(self, s) :
        l = 0
        dict_ = {}
        res = 0
        
        for r,char in enumerate(s):
            # duplicate
            if char in dict_ :
                
                # if duplicate inside l and r
                if dict_[char] >= l:  
                    l = dict_[char]+1
         
            dict_[char] = r
            res = max(res,r-l +1)
            
        return res

obj = Solution()
print(obj.lengthOfLongestSubstring("abcabcbb"))
# print(obj.lengthOfLongestSubstring("bbbbb"))
print(obj.lengthOfLongestSubstring("pwwkep"))

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0 
        res = curr = 0
        voc = dict() 
        for r in range(len(s)):
            if s[r] in voc and voc[s[r]]>=l:
                l = voc[s[r]]+1
                voc[s[r]] = r    
                curr = r-l +1
            else:
                voc[s[r]] = r
                curr+=1
            if curr>res:
                res = curr
        return res  

        

File: LeetCode/String/Medium/3. Longest Substring Without Repeating Characters/main.cpp
#include <iostream>
#include <string>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int start = 0;
        int max_len = 0;
        unordered_map<char, int> used;
        for (int i = 0; i < s.length(); i++)
        {
            // s[i] is current iterations char
            if (used.count(s[i]) && start <= used[s[i]])
            {
                cout << s[i] << i << endl;
                start = used[s[i]] + 1;
            }
            else
            {

                max_len = max(max_len, i - start + 1);
                cout << max_len << "\t" << i - start << endl;
            }
            used[s[i]] = i;
        }
        return max_len;
    }
};

int main()
{
    Solution obj;
    int ans = obj.lengthOfLongestSubstring("abcabcbb");
    cout << ans;
}